### PR TITLE
Xref spec lazy load

### PR DESCRIPTION
**Pull request by live1206**
- Create `XrefSpec` on demand, this will reduce the execution time of creating the internal xref map, which would save time for `docfx watch` with a single file.
- Only create `XrefSpec` for file with `Uid` defined on root level, ignore `XrefProperty` fields without root level `Uid`
- Fix nested method call with `ThreadStatic` field

### CODE REVIEW

File: src/docfx/build/xref/XrefMap.cs [hunk 1]
```diff
@@ -14,18 +14,18 @@ namespace Microsoft.Docs.Build
...
+        private readonly IReadOnlyDictionary<string, Lazy<XrefSpec>> _internalXrefMap;
         private readonly IReadOnlyDictionary<string, XrefSpec> _externalXrefMap;
 
-        public IEnumerable<XrefSpec> InternalReferences => _internalXrefMap.Values;
+        public IEnumerable<XrefSpec> InternalReferences => _internalXrefMap.Values.Select(v => v.Value);
 
         public XrefSpec Resolve(string uid)
         {
-            if (_internalXrefMap.TryGetValue(uid, out var xrefSpec))
+            if (_internalXrefMap.TryGetValue(uid, out var sepc))
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: internalSpec, externalSpec

File: src/docfx/build/xref/XrefMap.cs [hunk 2]
```diff
@@ -104,41 +107,44 @@ private static void Load(Context context, ConcurrentDictionary<string, Concurren
             }
         }
 
-        private static Dictionary<string, XrefSpec> HandleXrefConflicts(Context context, ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefsByUid)
+        private static SortedDictionary<string, Lazy<XrefSpec>> HandleXrefConflicts(Context context, ConcurrentDictionary<string, ConcurrentBag<Lazy<XrefSpec>>> xrefsByUid)
         {
-            var result = new List<XrefSpec>();
+            var result = new SortedDictionary<string, Lazy<XrefSpec>>();
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: are there any other places where we could use SortedDictionary?

File: src/docfx/build/xref/XrefMap.cs [hunk 3]
```diff
@@ -151,33 +157,32 @@ private static XrefSpec LoadSchemaDocument(List<Error> errors, JToken token, Doc
...
-                var xref = new XrefSpec
+                return (uid, new Lazy<XrefSpec>(() =>
                 {
-                    Uid = uid,
-                    Href = file.SitePath,
-                };
-                xref.ExtensionData.Merge(extensionData);
-                return xref;
+                    var (schemaErrors, content) = JsonUtility.ToObject(token, schema.Type, transform: AttributeTransformer.Transform(errors, file, null, extensionData));
+                    errors.AddRange(schemaErrors);
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: errors should not be manipulated here

File: src/docfx/lib/JsonUtility.cs [hunk 4]
```diff
@@ -176,7 +177,14 @@ public static (List<Error>, T) ToObject<T>(JToken token)
             try
             {
                 t_transform = transform;
```
Comment (yufeih) [inline src/docfx/lib/JsonUtility.cs]: should also push `t_transform`

File: src/docfx/build/xref/XrefMap.cs [hunk 5]
```diff
@@ -44,7 +46,7 @@ public static XrefMap Create(Context context, Docset docset)
                     map[sepc.Uid] = sepc;
                 }
             }
-            return new XrefMap(map, docset.Config.BuildInternalXrefMap ? CreateInternalXrefMap(context, docset.ScanScope) : new Dictionary<string, XrefSpec>());
+            return new XrefMap(map, docset.Config.BuildInternalXrefMap ? CreateInternalXrefMap(context, docset.ScanScope) : new Dictionary<string, Lazy<(List<Error>, XrefSpec)>>(), context);
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: >BuildInternalXrefMap [](start = 50, length = 20)

do we still need this switch after lazy loading?

File: src/docfx/build/xref/XrefMap.cs [hunk 6]
```diff
@@ -54,9 +56,9 @@ public void OutputXrefMap(Context context)
             context.WriteJson(models, "xrefmap.json");
         }
 
-        private static IReadOnlyDictionary<string, XrefSpec> CreateInternalXrefMap(Context context, IEnumerable<Document> files)
+        private static IReadOnlyDictionary<string, Lazy<(List<Error>, XrefSpec)>> CreateInternalXrefMap(Context context, IEnumerable<Document> files)
         {
-            ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefsByUid = new ConcurrentDictionary<string, ConcurrentBag<XrefSpec>>();
+            ConcurrentDictionary<string, ConcurrentBag<Lazy<(List<Error>, XrefSpec)>>> xrefsByUid = new ConcurrentDictionary<string, ConcurrentBag<Lazy<(List<Error>, XrefSpec)>>>();
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: >ConcurrentDictionary<string, ConcurrentBag<Lazy<(List<Error>, XrefSpec)>>> [](start = 12, length = 74)

var?

File: src/docfx/build/xref/XrefMap.cs [hunk 7]
```diff
@@ -151,33 +173,33 @@ private static XrefSpec LoadSchemaDocument(List<Error> errors, JToken token, Doc
...
 
-        private static void TryAddXref(ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefsByUid, XrefSpec spec)
+        private static void TryAddXref(ConcurrentDictionary<string, ConcurrentBag<Lazy<(List<Error>, XrefSpec)>>> xrefsByUid, string uid, Lazy<(List<Error>, XrefSpec)> spec)
         {
             if (spec is null)
             {
                 return;
             }
-            xrefsByUid.GetOrAdd(spec.Uid, _ => new ConcurrentBag<XrefSpec>()).Add(spec);
+            xrefsByUid.GetOrAdd(uid, _ => new ConcurrentBag<Lazy<(List<Error>, XrefSpec)>>()).Add(spec);
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: >GetOrAdd [](start = 23, length = 8)

just TryAdd?

File: docs/specs/xref.yml [hunk 8]
```diff
@@ -152,23 +152,30 @@ outputs:
...
+  docs/b.json: | 
+    {
+      "$schema": "https://raw.githubusercontent.com/dotnet/docfx/v3/schemas/TestData.json",
+      "xref": "a"
     }
 outputs:
-  docs/a.json:
+  docs/b.json: | 
+    {
+      "content": {}
```
Comment (yufeih) [inline docs/specs/xref.yml]: assert content does not have xref property using `{ "content": { "!xref": null } }`

**Pull request merged**